src/utils/src/arg_parser.rs

```diff
@@ -91,32 +91,52 @@ impl<'a> ArgParser<'a> {
     }
 }
 
+/// Type of argument.
+#[derive(Clone, Debug, PartialEq)]
+enum ArgumentType {
+    ValueArgument,
+    FlagArgument,
+}
+
 /// Stores the characteristics of the `name` command line argument.
 #[derive(Clone, Debug, PartialEq)]
 pub struct Argument<'a> {
     name: &'a str,
     required: bool,
     requires: Option<&'a str>,
-    takes_value: bool,
+    arg_type: ArgumentType,
     default_value: Option<Value>,
     help: Option<&'a str>,
     user_value: Option<Value>,
 }
```

Comment:
Changing `takes_value: bool` with `arg_type: ArgumentType` does provide a little extra clarity, but just as before you still have a invisible/assumed connection between the two fields in the struct `arg_type: ArgumentType` and `user_value: Option<Value>` whose values implicitly depend on one another.

If I understand correctly, with this version the compiler allows `user_value == Some(Value)` while `arg_type == ArgumentType::FlagArgument`, while the code logic considers this case to be invalid.

One of the powerful tools in rust to remove these error-prone implicit dependencies and replace them with explicit compile-time validated dependencies is using a single enum that defines only valid states and their possible data:

```rust
enum ArgumentType {
    ValueArgument(Value),
    FlagArgument,
}
```
If this field can be None you can wrap this enum in an `Option`, or even better code is to add another variant to it `NotPresent` or smth.

Reply:
While looking a bit more at the code, there are quite a few methods that don't make sense for both `ValueArgument` and `FlagArgument`.

You can call the following methods on the `ValueArgument`:
- required
- requires
- default_value
- help

But on the `FlagArgument` only the following methods make sense:
- requires
- help

An `Argument` trait might still make sense even if we duplicate a few lines of code in both `ValueArgument` and `FlagArgument`.

Looks like we are very deep in the rabbit hole :))